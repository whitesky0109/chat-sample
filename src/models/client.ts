import { UserPayload } from 'public/store/actions/userAction';
import { RoomPayload } from 'public/store/actions/roomAction';

export interface IMessage {
  roomId?: string;
  type: string;
  message: File | string;
}

export interface StoreState {
  user: UserPayload;
  room: RoomPayload;
}
