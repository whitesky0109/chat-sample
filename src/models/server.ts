export interface IMigration {
  migration_id: string;
  action: string;
  args: any[];
}

export interface IRepository {
  migrations(): IMigration[];
}

export interface IService {
  init(): Promise<any>;
}

export interface IMigrationLog {
  id: number;
  migration_id: string;
  action: string;
  success: boolean;
  error: string;
  timestamp: Date;
}

export interface IMigrationLogs {
  // tslint:disable-next-line: variable-name
  [migration_id: string]: IMigrationLog;
}

export interface IInvite {
  roomId: string;
  users: string[];
}

export interface IMessage {
  userId: string;
  type: string;
  message: ArrayBuffer | string;
  date: Date;
  isMy: boolean;
}

export interface IUser {
  [userId: string]: string | null;
}

export interface IRoomInfo {
  name: string;
  users: IUser;
  creator: string;
}

export interface IRoom {
  [roomId: string]: IRoomInfo;
}
