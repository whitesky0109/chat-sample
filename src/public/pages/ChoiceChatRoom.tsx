import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import React, { Component, ReactNode, ChangeEvent, KeyboardEvent } from 'react';

// models
import { StoreState } from 'models/client';

// components
import RoomList from '../components/RoomList';

// redux
import { setRoomId, clearMessage } from '../store/actions/roomAction';

// etc
import { reqAddRoom, reqRooms, reqEnterRoom, reqUserLogout } from '../controllers/socket';

export interface Props extends RouteComponentProps {
  room?: any;
  roomId?: string;
  user?: string;

  setRoomId: typeof setRoomId;
  clearMessage: typeof clearMessage;
}

export interface States {
  value: string;
}

export class ChoiceChatRoom extends Component<Props, States> {
  state: States = {
    value: '',
  };

  componentWillMount() {
    const { history, user, setRoomId, clearMessage } = this.props;
    if (!user) {
      history.push('/');
      return;
    }

    setRoomId();
    clearMessage();
    reqRooms();
  }

  shouldComponentUpdate(nextProps: Props) {
    const { history, roomId } = this.props;
    const { roomId: nextRoomId, user } = nextProps;

    if (!user) {
      history.push('/');
      return false;
    }

    if (nextRoomId) {
      if (roomId) {
        clearMessage();
      }
      history.push(`/room/${nextRoomId}`);
      return false;
    }

    return true;
  }

  addRoom = () => {
    const { value } = this.state;
    if (value) {
      reqAddRoom(value);
    }
    this.setState({ value: '' });
  }

  enterRoom = (roomId: string) => {
    reqEnterRoom(roomId);
  }

  onInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;

    this.setState({ value });
  }

  onInputKeyPress = (e: KeyboardEvent<HTMLInputElement>) => {

    if (e.keyCode === 13 && !e.shiftKey) {
      e.preventDefault();
      this.addRoom();
    }
  }

  logout = () => {
    reqUserLogout();
  }

  render(): ReactNode {
    const { room, user } = this.props;
    const { value } = this.state;

    if (!room || !user) {
      return (
        <> </>
      );
    }

    return (
      <div className="container">
        <div className="header">
          <div className="title">
            <div className="user-icon">
              <i className="fa fa fa-weixin" />
            </div>
            <div className="about">
              <div className="chat-with">{user}</div>
            </div>
          </div>

          <div className="btn-group">
            <button className="btn" onClick={this.logout} title="로그 아웃">
              <i className="fa fa-share-square-o" />
            </button>
          </div>
        </div>

        <div className="people-list" id="people-list">

          <div className="room-input">
            <input
              className="text"
              placeholder="Type Chat room name"
              value={value}
              onChange={this.onInputChange}
              onKeyDown={this.onInputKeyPress}
            />
            <button onClick={this.addRoom} title="방 만들기"><i className="fa fa-plus" /></button>
          </div>

          <RoomList rooms={room} onRoomClick={this.enterRoom} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = {
  setRoomId,
  clearMessage,
};

const mapStateToProps = (state: StoreState) => {
  const { room, user } = state;
  return {
    room: room.room,
    user: user.user,
    roomId: room.roomId,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ChoiceChatRoom);
