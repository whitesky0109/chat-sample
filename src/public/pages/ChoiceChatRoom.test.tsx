import React from 'react';
import { ReactWrapper, mount } from 'enzyme';
import { StoreState } from 'models/client';
import createMockStore from 'redux-mock-store';
import ChoiceChatRoom, { Props } from './ChoiceChatRoom';
import { Provider } from 'react-redux';

describe('ChoiceChatRoom', () => {
  let component: ReactWrapper;
  let ccr: ReactWrapper;

  const mock: any = jest.fn();
  const props: Props = {
    match: mock,
    location: mock,
    history: {
      goBack: mock,
    } as any,

    setRoomId: mock,
    clearMessage: mock,
  };

  const generateFakeStore = (initState: StoreState) => {
    const mockStore = createMockStore();
    return mockStore(initState);
  };

  it('render', () => {
    const store = generateFakeStore({
      user: {
        user: 'userId',
        users: {
          ['userId']: 'sockid',
        },
      },
      room: { },
    });
    component = mount(<Provider store={store}>
      <ChoiceChatRoom {...props} />
    </Provider>);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('find ChoiceChatRoom component', () => {
    ccr = component.find('ChoiceChatRoom');
    expect(ccr).toHaveLength(1);
  });

});
