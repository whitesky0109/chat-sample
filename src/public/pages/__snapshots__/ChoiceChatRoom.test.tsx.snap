// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoiceChatRoom matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ChoiceChatRoom)
    clearMessage={[MockFunction]}
    history={
      Object {
        "goBack": [MockFunction],
      }
    }
    location={[MockFunction]}
    match={[MockFunction]}
    setRoomId={[MockFunction]}
  >
    <ChoiceChatRoom
      clearMessage={[Function]}
      history={
        Object {
          "goBack": [MockFunction],
        }
      }
      location={[MockFunction]}
      match={[MockFunction]}
      setRoomId={[Function]}
      user="userId"
    >
       
    </ChoiceChatRoom>
  </Connect(ChoiceChatRoom)>
</Provider>
`;
